### ISSUE TITLE
[Question] - OnModelCreating and Global Query Filters

### ISSUE THREAD

**Issue by dazinator** [dotnet/efcore#10322]
I am using `EF Core 20` in an Xamarin app (monodroid80) targeting `netstandard20`.

Multiple users can use the app, as they may share the same device between them. Imagine a team, where one user has the device one day, then another user the next day etc.

However I need to keep each user's data completely seperate.

My current approach is to have a single database, but treat each user as a tenant, and I have added `TenantID` to all tables (i.e the base entity), and now I am looking to use a global query filter to automatically append TenantID filter to all queries (i.e `where TenantID = [current user id]`)

I am using Prism, with an Autofac Container Adaptor, with the `IServiceCollection` adaptor! :-)
Basically I can register services as normal from an `IServiceCollection` on app startup - so I register DbContext in the normal way i.e  `services.AddDbContext<>`

However because this is not a web app, there is never a "scope" created or disposed. (i.e there is no Per-Request scope!). This basically means I am seeing a single `DbContext` in use for the life of the app.

My question is, if my app is using a single instance of the `DbContext`, and inside it's OnModelCreating is this:

```            
            var tenantId = _currentTenant.Id;          
            builder.Entity<T>().HasQueryFilter(e => e.TenantId == tenantId && !e.IsDeleted);

```

If _currentTenant.Id is then changed at runtime, will future queries automatically use the new value?
It's not clear to me exactly when `OnModelCreating` will get fired - is it once per instance of DbContext? Is it once per AppContext? Do we have to be wary of the lambda  `capturing` stale values?

I am also wondering whether the approach I am taking (i.e no scope, and using a single instance of DbContext app wide) is the recommended approach, or perhaps it would be better to have an instance of `DbContext` "per tenant" using a per-tenant child container approach. 

Per Tenant Child Container approach could work in that, when the app starts up, DbContext **would not** be registered into the main container. When a User signs in - a Child Container would be created (and DbContext registered into that) for that tenant, and then the main application container would be swapped out for the current tenant child container (similar to what asp.net core container middleware does per request) for the duration the user was signed in. When the user signs out, the child container could be disposed of and the main container restored again. However this wouldn't work if `OnModelCreating` is only called once per `AppContext` and is not somehow in anyway scoped to the container registration.

**Comment by smitpatel** [comment]
```C#
 var tenantId = _currentTenant.Id;          
            builder.Entity<T>().HasQueryFilter(e => e.TenantId == tenantId && !e.IsDeleted);
```
By using local variable in filter you effectively broke the linkage with context. So we no longer know which part comes from DbContext.

If you configure your filter like this
`builder.Entity<T>().HasQueryFilter(e => e.TenantId == _currentTenant.Id && !e.IsDeleted);`
then it will inject value of _currentTenant at runtime from db context instance. Assuming _currentTenant is field/property on your DbContext deriving class.

**Comment by dazinator** [comment]
@smitpatel - thanks for responding. This linkage to `DbContext` instance members seems a bit magical if its what I am imagining it to be. There is no mention of this in the docs: https://docs.microsoft.com/en-us/ef/core/querying/filters is it worth adding something - unless I am misunderstanding something which is always possible.

I was suspecting to see an API similar in nature to the following when this feature was released:

```
builder.Entity<T>().HasQueryFilter(e => e.TenantId == Ef.GlobalParam<int>("foo"));

```

With the actual value having to be added to some dictionary or something somewhere so that it could be supplied at runtime.

```
dbContext.GlobalParamaters.Add<int>("foo", 1);
```

That has the benefit that it is obvious where the state is, and can be changed in the dictionary fairly easy. But this magic API has left me guessing a few things!

**Comment by anpete** [comment]
@dazinator It is called out in the tip that starts with "Note the use of a DbContext instance level field..."

This is the first version of these docs, though, and so surely there are lots of improvements that could be made. Did I mention we love getting PRs on the docs? 😀

**Comment by paillave** [comment]
Does it mean global query filters for multitenant cannot work if they are configured in IEntityTypeConfiguration?

**Comment by ajcvickers** [comment]
@paillave No. It means that if the filter is using a property of the context which can change, then it must be referenced explicitly in the expression tree to be picked up by EF.

**Comment by paillave** [comment]
that exactly where lie my problem: how to make reference to the context in the  class that implements IEntityTypeConfiguration? Is there anywhere where I can find an example?

**Comment by ajcvickers** [comment]
@paillave You need to pass the context instance into the configuration class when it is created in `OnModelCreating`.

**Comment by paillave** [comment]
2 more questions then: 
- EF will automatically detect that the global query filter uses a value that has the same type than the context and will replace it in the expression by the actual context?
- that means I can't use `ApplyConfigurationsFromAssembly` anymore?

**Comment by ajcvickers** [comment]
@dazinator 1. Essentially, yes. 2. Yes, but that's just sugar that can easily be moved to application code and changed to accept a context instance.

**Comment by paillave** [comment]
If 1. is yes, would it actually work if I use a context that is simply declared and not even instantiated?

**Comment by ajcvickers** [comment]
@paillave Not by design, but we currently have a bug that means this might work. It would not be wise to rely on this continuing to work.

**Comment by paillave** [comment]
and what if I provide a dummy instance of a context in the `IEntityTypeConfiguration` (like one that is plugged to in memory db). Would it be "wiser"? :)

**Comment by ajcvickers** [comment]
@paillave It really needs to be the same context instance that you are using to build the model.

**Comment by paillave** [comment]
Actually I initially did exactly what you advice me: from `OnModelCreating` I transmitted the context to the constructor of `IEntityTypeConfiguration` so that I can setup the global query filter in there. But EF actually ignored the global query filter. It even seemed that `OnModelCreating` was not even called at run time (I admit that I might have mistaken in the way I tested to the conclusion that `OnModelCreating` may not be called at runtime). The only thing that worked for me is the global query filter with the context, making reference to a local field of it.
 , but I faced an issue. Initially, , but at run time,  is not called.
